### ISSUE TITLE
Group entities (eg. by areas of interes, by collections and subcollections, etc) [ER Diagram]

### ISSUE THREAD

**Issue by st-trade** [mermaid-js/mermaid#3225]

**Comment by IloBe** [comment]
Great enhancement, I would like to have that too ...

**Comment by nitsujri** [comment]
In my search for this feature, I found [d2](https://github.com/terrastruct/d2) which I am using. Large entity graphs get very hard to read without some logical

Would love Mermaid to be able to do this so I can stay in one DSL!

**Comment by hdsheena** [comment]
I came looking for this and didn't find it at first because it's missing the Entity Relationship tag/label.

**Comment by psimona** [comment]
Upvote the feature. Would 100% use

**Comment by logicguy1** [comment]
I started using namespaces for this, seems to give a similar result, but i don't think it is possible to nest namespaces.
Would love to group contexts inside modules.

**Comment by ryan-ajility** [comment]
This feature is a must.

If you're building an application that is large enough to be valuable and you don't want a ball of mud, separating the code into modules is necessary.

These modules need to be represented in the ER diagram. Otherwise, an entire layer of context is hidden from developers.

**Comment by upkeihas** [comment]
Yes please, I would also like to group my entities. This is a good feature.

**Comment by npiguet** [comment]
I also need this.

Currently, the best workaround is to use flow diagrams and sub-graphs. But flow diagrams are pretty bad at displaying the cardinality of the relationship between entities.

Another workaround is to color nodes according to which package/module/group they belong to, but:
1. It is tedious to maintain
2. You cannot label the groups

**Comment by chrispy-snps** [comment]
We would love this too!

**Comment by isalew** [comment]
I don't know how to increase the prioritization of this other than to say ditto to the many voices on this enhancement request. At this point, can't somebody just use AI to put a pull request together? 😜🤖

**Comment by aloisklink** [comment]
I'm a big fan of this idea, and it should be relatively easily to implement now that ER diagrams is re-using much of the same layout/rendering code as flowcharts.

The main issue would be designing the syntax for it. I would have loved to use `namespace {}` like in class diagrams: https://mermaid.js.org/syntax/classDiagram.html#namespace-labels-v11-15-0 but somebody might have made an existing ER diagram that uses that syntax to declare an entity named `namespace`, e.g. something like the following is a valid existing diagram:

```mermaid-example
erDiagram
    namespace Identity["Identity & Access"] {
        USER
        ROLE
    }
```

```mermaid
erDiagram
    namespace Identity["Identity & Access"] {
        USER
        ROLE
    }
```

We might need to do use some unusual characters like `group("My label for the group") { entity1 entity2 entity3 }` (since `()` is currently not allowed like this) to make sure this doesn't break any existing diagrams.

---

Feel free to 👍 this issue if you want to see it prioritized! If you comment on this, you'd just ping everybody in the issue unnecessarily.

---

> I started using namespaces and class diagrams for this, seems to give a similar result, but i don't think it is possible to nest namespaces.

_From @logicguy1 in https://github.com/mermaid-js/mermaid/issues/3225#issuecomment-2717073250_

Class diagrams do now support nested namespaces, as of Mermaid v11.15.0! Check out https://github.com/mermaid-js/mermaid/pull/7604 and https://mermaid.js.org/syntax/classDiagram.html#nested-namespaces-v11-15-0.
